packages/teams-js/src/internal/videoUtils.ts

```diff
@@ -78,36 +135,186 @@ function createProcessedStreamGenerator(
+          originalFrame,
+        );
+        const frameProcessedByApp = await this.videoFrameHandler({ videoFrame, audioInferenceResult });
+        // the current typescript version(4.6.4) dosn't support webcodecs API fully, we have to do type conversion here.
+        const processedFrame = new VideoFrame(frameProcessedByApp as unknown as CanvasImageSource, {
+          // we need the timestamp to be unchanged from the oirginal frame, so we explicitly set it here.
+          timestamp: timestamp,
+        });
+        controller.enqueue(processedFrame);
+        videoFrame.close();
+        originalFrame.close();
+        (frameProcessedByApp as VideoFrame).close();
+      } catch (error) {
+        originalFrame.close();
+        this.notifyError(error);
+      }
+    } else {
+      this.notifyError('timestamp of the original video frame is null');
+    }
+  };
+
+  /**
+   * @hidden
+   * Extract video frame and metadata from the given texture.
+   * The given texure should be in NV12 format and the layout of the texture should be:
+   * | Texture layout        |
+   * | :---                  |
+   * | Header                |
+   * | Real video frame data |
+   * | Metadata              |
+   *
+   * The header data is in the first two rows with the following format:
+   * | oneTextureLayoutId | version | frameRowOffset | frameFormat | frameWidth | frameHeight | multiStreamHeaderRowOffset | multiStreamCount | ...   |
+   * |    :---:           | :---:   | :---:          |  :---:      |  :---:     |  :---:      |  :---:                     |  :---:           | :---: |
+   * | 4 bytes            | 4 bytes | 4 bytes        | 4 bytes     | 4 bytes    | 4 bytes     | 4 bytes                    | 4 bytes          | ...   |
+   *
+   * After header, it comes with the real video frame data.
+   * At the end of the texture, it comes with the metadata. The metadata section can contain multiple types of metadata.
+   * Each type of metadata is called a stream. The section is in the following format:
+   * | stream1.id | stream1.dataOffset | stream1.dataSize | stream2.id | stream2.dataOffset | stream2.dataSize | ... | stream1.data | stream2.data | ... |
+   * | :---:      | :---:              | :---:            |  :---:     |  :---:             |  :---:           |:---:|  :---:       | :---:        |:---:|
+   * | 4 bytes    | 4 bytes            | 4 bytes          | 4 bytes    | 4 bytes            | 4 bytes          | ... | ...          | ...          | ... |
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  private extractVideoFrameAndMetadata = async (
+    texture: VideoFrame,
+  ): Promise<{ videoFrame: VideoFrame; metadata: { audioInferenceResult?: Uint8Array } }> => {
+    if (inServerSideRenderingEnvironment()) {
+      throw errorNotSupportedOnPlatform;
+    }
+
+    if (texture.format !== 'NV12') {
+      this.notifyError('Unsupported video frame format');
```

Comment:
Please only define this string once, using either a const string or a const Error